File: backend/tripoptimizer/api/schemas.py
"""Pydantic models for the HTTP boundary.

Request validation here enforces the MVP guardrails the pure core deliberately
omits (max cities, flex cap). Response models serialize the core's frozen
dataclasses to JSON via ``from_core`` converters, keeping the core Pydantic-free.
"""

from __future__ import annotations

from datetime import date

from pydantic import BaseModel, Field, model_validator

from tripoptimizer.core.optimizer.models import Itinerary, TripResult

MAX_CITIES = 8
MAX_FLEX_DAYS = 7


def aggregate_data_source(itinerary: Itinerary) -> str:
    """cached / synthetic when uniform across legs, else 'mixed'."""
    sources = {leg.source for leg in itinerary.legs}
    return next(iter(sources)) if len(sources) == 1 else "mixed"


class TripRequestSchema(BaseModel):
    cities: list[str] = Field(..., min_length=1, max_length=MAX_CITIES)
    days_per_city: dict[str, int]
    origin_airport: str
    return_airport: str
    start_date: date
    flex_days: int = Field(default=3, ge=0, le=MAX_FLEX_DAYS)

    @model_validator(mode="after")
    def _days_cover_cities(self) -> "TripRequestSchema":
        missing = [c for c in self.cities if c not in self.days_per_city]
        if missing:
            raise ValueError(f"days_per_city is missing entries for: {sorted(missing)}")
        for city in self.cities:
            if self.days_per_city[city] <= 0:
                raise ValueError(f"days_per_city[{city!r}] must be > 0")
        return self


class LegSchema(BaseModel):
    origin: str
    destination: str
    fly_date: date
    price: float
    source: str


class ItinerarySchema(BaseModel):
    order: list[str]
    start_offset: int
    legs: list[LegSchema]
    total: float

    @classmethod
    def from_core(cls, itinerary: Itinerary) -> "ItinerarySchema":
        return cls(
            order=list(itinerary.order),
            start_offset=itinerary.start_offset,
            legs=[
                LegSchema(
                    origin=leg.origin,
                    destination=leg.destination,
                    fly_date=leg.fly_date,
                    price=leg.price,
                    source=leg.source,
                )
                for leg in itinerary.legs
            ],
            total=itinerary.total,
        )


class TripResultSchema(BaseModel):
    best: ItinerarySchema
    alternatives: list[ItinerarySchema]
    data_source: str
    snapshot_date: date | None = None

    @classmethod
    def from_core(
        cls,
        result: TripResult,
        *,
        data_source: str | None = None,
        snapshot_date: date | None = None,
    ) -> "TripResultSchema":
        return cls(
            best=ItinerarySchema.from_core(result.best),
            alternatives=[ItinerarySchema.from_core(it) for it in result.alternatives],
            data_source=data_source or aggregate_data_source(result.best),
            snapshot_date=snapshot_date,
        )


class AirportSchema(BaseModel):
    iata: str
    name: str
    city: str
    country: str
    lat: float
    lon: float

File: frontend/src/lib/schemas.ts
import { z } from "zod";

export const AirportSchema = z.object({
  iata: z.string(),
  name: z.string(),
  city: z.string(),
  country: z.string(),
  lat: z.number(),
  lon: z.number(),
});
export type Airport = z.infer<typeof AirportSchema>;

export const LegSchema = z.object({
  origin: z.string(),
  destination: z.string(),
  fly_date: z.string(),
  price: z.number(),
  source: z.string(),
});
export type Leg = z.infer<typeof LegSchema>;

export const ItinerarySchema = z.object({
  order: z.array(z.string()),
  start_offset: z.number(),
  legs: z.array(LegSchema),
  total: z.number(),
});
export type Itinerary = z.infer<typeof ItinerarySchema>;

// A full result: every leg is priced from real data (`cached`; `mixed` if sources differ).
export const TripResultSchema = z.object({
  status: z.literal("ok"),
  best: ItinerarySchema,
  alternatives: z.array(ItinerarySchema),
  data_source: z.enum(["cached", "mixed"]),
  snapshot_date: z.string().nullable(),
});
export type TripResult = z.infer<typeof TripResultSchema>;

// No fully-real itinerary exists: these (origin, destination) pairs had no real fare in
// the window. Shown honestly instead of a fabricated price (real-or-nothing).
export const IncompleteTripSchema = z.object({
  status: z.literal("incomplete"),
  missing_routes: z.array(z.tuple([z.string(), z.string()])),
  snapshot_date: z.string().nullable(),
});
export type IncompleteTrip = z.infer<typeof IncompleteTripSchema>;

// The /optimize response is one of the two shapes, discriminated by `status`.
export const OptimizeResponseSchema = z.discriminatedUnion("status", [
  TripResultSchema,
  IncompleteTripSchema,
]);
export type OptimizeResponse = z.infer<typeof OptimizeResponseSchema>;

// Client-side request model. Mirrors the backend guardrails so requests are valid before sending.
export const TripRequestSchema = z
  .object({
    origin_airport: z.string().length(3),
    return_airport: z.string().length(3),
    cities: z.array(z.string().length(3)).min(1).max(8),
    days_per_city: z.record(z.string(), z.number().int().positive()),
    start_date: z.string().regex(/^\d{4}-\d{2}-\d{2}$/),
    flex_days: z.number().int().min(0).max(7),
  })
  .refine((r) => r.cities.every((c) => r.days_per_city[c] > 0), {
    message: "Every city needs days > 0",
    path: ["days_per_city"],
  });
export type TripRequest = z.infer<typeof TripRequestSchema>;
